lib/Sema/TypeCheckConcurrency.h

```diff
@@ -580,4 +591,49 @@ bool diagnoseApplyArgSendability(
 
 } // end namespace swift
 
+namespace llvm {
+
+template <>
+struct DenseMapInfo<swift::ReferencedActor::Kind> {
+  using RefActorKind = swift::ReferencedActor::Kind;
+
+  static RefActorKind getEmptyKey() {
+   return RefActorKind::NonIsolatedContext;
+  }
+
+  static RefActorKind getTombstoneKey() {
+   return RefActorKind::NonIsolatedContext;
+  }
+
+  static unsigned getHashValue(RefActorKind Val) {
+   return static_cast<unsigned>(Val);
+  }
+
+  static bool isEqual(const RefActorKind &LHS, const RefActorKind &RHS) {
+   return LHS == RHS;
+  }
+ };
+
+  template <>
+  struct DenseMapInfo<swift::ActorIsolation> {
+    using RefActor = swift::ActorIsolation;
+
+    static RefActor getEmptyKey() {
+      return RefActor(swift::ActorIsolation::Kind::Unspecified);
+    }
+
+    static RefActor getTombstoneKey() {
+     return RefActor(swift::ActorIsolation::Kind::Unspecified);
+    }
+
+    static unsigned getHashValue(RefActor Val) {
+     return static_cast<unsigned>(Val.getKind());
```

Comment:
This is going to be every inefficient to store in a map because different actor isolations with the same kind would just form a list so to find a particular item would mean an extensive linear scan. I think the hashing needs to account for the underlying declaration and preconcurrency bit at least.

Reply:
Looks like ActorIsolation already has a friend `hash_value` function too that could be used here.

Reply:
Hmm, the ActorIsolation hash will not be used on its own though when serving as a map key. 

When used as a key for the reference errors, it is paired with the ReferencedActor::Kind: 

```
    using ActorRefKindPair = std::pair<ReferencedActor::Kind, ActorIsolation>;
    llvm::DenseMap<ActorRefKindPair, DiagnosticList> refErrors;
```

This should be ok because we want to have a map that groups all diagnostics with the same ReferencedActor Kind and ActorIsolation kind.

Reply:
The pair would take hash of each element and combine so you don’t have to worry about that.